File: python/tests/rtc/ds3231_test_unit.py
"""Unit test for the DS3231 — runs without hardware using the I2C mock.

Verifies BCD conversion, 24-hour enforcement, alarm match-mode encoding,
and interrupt status masking by observing what the driver writes to the
mock connection.
"""

from periph.connection.i2c_mock import I2CConnectionMock
from periph.chips.rtc.ds3231 import (
    DS3231Minimal, DS3231Full,
    ALARM1_MATCH_HOURS_MINUTES_SECONDS, ALARM1_MATCH_DAY_HOURS_MINUTES_SECONDS,
    ALARM2_EVERY_MINUTE,
    SOURCE_ALARM1, SOURCE_ALARM2,
)

passed = 0
failed = 0


def check_true(label, condition):
    global passed, failed
    if condition:
        print('PASS', label)
        passed += 1
    else:
        print('FAIL', label)
        failed += 1


# get_datetime: 2026-09-22 Tuesday 14:30:45, 24-hour mode.
mock = I2CConnectionMock()
mock.set_register(DS3231Minimal._REG_SECONDS, 0x45, 0x30, 0x14, 0x02, 0x22, 0x09, 0x26)
rtc = DS3231Minimal(mock)
dt = rtc.get_datetime()
check_true('get_datetime_time', dt[4:7] == (14, 30, 45))
check_true('get_datetime_date', dt[0:4] == (2026, 9, 22, 2))

# get_datetime: 12-hour mode, 11 PM -> 23:00.
mock2 = I2CConnectionMock()
mock2.set_register(DS3231Minimal._REG_SECONDS, 0x00, 0x00, 0x71, 0x01, 0x01, 0x01, 0x25)
rtc2 = DS3231Minimal(mock2)
check_true('get_datetime_12h_pm_decode', rtc2.get_datetime()[4] == 23)

# set_datetime: forces 24-hour mode and clears OSF.
mock3 = I2CConnectionMock()
mock3.set_register(DS3231Minimal._REG_STATUS, 0x80)  # OSF set
rtc3 = DS3231Minimal(mock3)
rtc3.set_datetime(2026, 9, 22, 2, 14, 30, 45)
hour_ok = any(len(w) >= 4 and w[0] == DS3231Minimal._REG_SECONDS and w[3] == 0x14 for w in mock3.writes)
osf_cleared = any(len(w) == 2 and w[0] == DS3231Minimal._REG_STATUS and w[1] == 0x00 for w in mock3.writes)
check_true('set_datetime_writes_24h_bcd_hour', hour_ok)
check_true('set_datetime_clears_osf', osf_cleared)

# read_temperature: +25.25 C (MSB=0x19, LSB=0x40).
mock4 = I2CConnectionMock()
mock4.set_register(DS3231Minimal._REG_TEMP_MSB, 0x19, 0x40)
rtc4 = DS3231Minimal(mock4)
check_true('read_temperature_positive', abs(rtc4.read_temperature() - 25.25) < 1e-9)

# read_temperature: negative, -9.75 C (MSB=0xF6 = -10, frac=0.25).
mock5 = I2CConnectionMock()
mock5.set_register(DS3231Minimal._REG_TEMP_MSB, 0xF6, 0x40)
rtc5 = DS3231Minimal(mock5)
check_true('read_temperature_negative', abs(rtc5.read_temperature() - (-9.75)) < 1e-9)

# Alarm1 match-mode round trip: MATCH_HOURS_MINUTES_SECONDS.
mock6 = I2CConnectionMock()
rtc6 = DS3231Full(mock6)
rtc6.set_alarm1(15, 30, 9, 0, False, ALARM1_MATCH_HOURS_MINUTES_SECONDS)
a1 = rtc6.get_alarm1()
check_true('alarm1_match_mode_roundtrip', a1['match_mode'] == ALARM1_MATCH_HOURS_MINUTES_SECONDS)
check_true('alarm1_fields_roundtrip', (a1['second'], a1['minute'], a1['hour']) == (15, 30, 9))

# Alarm1 match-mode round trip: MATCH_DAY (day-of-week).
mock7 = I2CConnectionMock()
rtc7 = DS3231Full(mock7)
rtc7.set_alarm1(0, 0, 0, 3, True, ALARM1_MATCH_DAY_HOURS_MINUTES_SECONDS)
a1b = rtc7.get_alarm1()
check_true('alarm1_day_mode_roundtrip', a1b['match_mode'] == ALARM1_MATCH_DAY_HOURS_MINUTES_SECONDS)
check_true('alarm1_day_fields_roundtrip', a1b['is_day_of_week'] and a1b['day_or_date'] == 3)

# Alarm2 match-mode round trip: EVERY_MINUTE.
mock8 = I2CConnectionMock()
rtc8 = DS3231Full(mock8)
rtc8.set_alarm2(0, 0, 0, False, ALARM2_EVERY_MINUTE)
check_true('alarm2_every_minute_roundtrip', rtc8.get_alarm2()['match_mode'] == ALARM2_EVERY_MINUTE)

# Interrupt status masking: poll_interrupt clears A1F/A2F, leaves OSF/EN32kHz.
mock9 = I2CConnectionMock()
mock9.set_register(DS3231Minimal._REG_STATUS, 0x80 | 0x08 | 0x01 | 0x02)
rtc9 = DS3231Full(mock9)
status = rtc9.poll_interrupt()
check_true('poll_interrupt_reports_alarm1', bool(status & SOURCE_ALARM1))
check_true('poll_interrupt_reports_alarm2', bool(status & SOURCE_ALARM2))
after = [w[1] for w in mock9.writes if len(w) == 2 and w[0] == DS3231Minimal._REG_STATUS][-1]
check_true('poll_interrupt_clears_alarm_flags', (after & 0x01) == 0 and (after & 0x02) == 0)
check_true('poll_interrupt_preserves_osf_en32khz', (after & 0x80) != 0 and (after & 0x08) != 0)

# enableInterrupt sets INTCN + A1IE/A2IE.
mock10 = I2CConnectionMock()
rtc10 = DS3231Full(mock10)
rtc10.enable_interrupt(SOURCE_ALARM1 | SOURCE_ALARM2)
ctrl = [w[1] for w in mock10.writes if len(w) == 2 and w[0] == DS3231Minimal._REG_CONTROL][-1]
check_true('enable_interrupt_sets_intcn_and_both_ie_bits', (ctrl & 0x04) and (ctrl & 0x01) and (ctrl & 0x02))

# Aging offset: signed round trip.
mock11 = I2CConnectionMock()
mock11.set_register(DS3231Minimal._REG_AGING_OFFSET, 0xFB)  # -5
rtc11 = DS3231Full(mock11)
check_true('aging_offset_signed_read', rtc11.get_aging_offset() == -5)
rtc11.set_aging_offset(-100)
raw = [w[1] for w in mock11.writes if len(w) == 2 and w[0] == DS3231Minimal._REG_AGING_OFFSET][-1]
check_true('aging_offset_signed_write', raw == 0x9C)  # -100 & 0xFF

print('===DONE: %d passed, %d failed===' % (passed, failed))
if failed:
    raise SystemExit(1)

File: nodejs/tests/rtc/ds3231_test_unit.js
'use strict';
const { I2CConnectionMock } = require('../../packages/periph/src/connection/i2c_mock');
const { DS3231Minimal, DS3231Full } = require('../../packages/periph/src/chips/rtc/ds3231');

const _REG_SECONDS         = 0x00;
const _REG_MINUTES         = 0x01;
const _REG_HOURS           = 0x02;
const _REG_DAY             = 0x03;
const _REG_DATE            = 0x04;
const _REG_MONTH_CENTURY   = 0x05;
const _REG_YEAR            = 0x06;
const _REG_ALARM1_SECONDS  = 0x07;
const _REG_ALARM1_DAY_DATE = 0x0A;
const _REG_ALARM2_MINUTES  = 0x0B;
const _REG_ALARM2_DAY_DATE = 0x0D;
const _REG_CONTROL         = 0x0E;
const _REG_CONTROL_STATUS  = 0x0F;
const _REG_AGING_OFFSET    = 0x10;
const _REG_TEMP_MSB        = 0x11;
const _REG_TEMP_LSB        = 0x12;

let passed = 0;
let failed = 0;

function checkTrue(label, condition) {
    if (condition) { console.log('PASS', label); passed++; }
    else { console.log('FAIL', label); failed++; }
}

function checkEq(label, got, expected) {
    if (got === expected) { console.log('PASS', label); passed++; }
    else { console.log(`FAIL ${label}: got ${got}, expected ${expected}`); failed++; }
}

async function main() {
    // -- init(): plain presence read, no writes --------------------------
    {
        const conn = new I2CConnectionMock();
        const rtc = new DS3231Minimal(conn);
        await rtc.init();
        // The mock logs the register-address byte of a writeRead() as a "write"
        // (it's a real bus write before the repeated-start read), so one entry
        // here is the presence probe itself, not a register-value write.
        checkEq('init_probe_is_address_byte_only', conn.writes.length, 1);
        checkTrue('init_reads_control_reg', conn.registers.size === 0); // read-only probe; nothing preloaded, nothing changed
    }

    // -- setDatetime() / getDatetime() round trip, and OSF clearing ------
    {
        const conn = new I2CConnectionMock();
        conn.setRegister(_REG_CONTROL_STATUS, [0x88]); // OSF=1, BSY=1
        const rtc = new DS3231Minimal(conn);
        await rtc.setDatetime(2026, 9, 22, 2, 14, 30, 0);

        checkEq('set_datetime_seconds', conn.registers.get(_REG_SECONDS), 0x00);
        checkEq('set_datetime_minutes', conn.registers.get(_REG_MINUTES), 0x30);
        checkEq('set_datetime_hours', conn.registers.get(_REG_HOURS), 0x14);
        checkEq('set_datetime_weekday', conn.registers.get(_REG_DAY), 0x02);
        checkEq('set_datetime_date', conn.registers.get(_REG_DATE), 0x22);
        checkEq('set_datetime_month', conn.registers.get(_REG_MONTH_CENTURY), 0x09);
        checkEq('set_datetime_year', conn.registers.get(_REG_YEAR), 0x26);
        checkEq('set_datetime_clears_osf_only', conn.registers.get(_REG_CONTROL_STATUS), 0x08);

        const dt = await rtc.getDatetime();
        checkEq('get_datetime_year', dt.year, 2026);
        checkEq('get_datetime_month', dt.month, 9);
        checkEq('get_datetime_day', dt.day, 22);
        checkEq('get_datetime_weekday', dt.weekday, 2);
        checkEq('get_datetime_hour', dt.hour, 14);
        checkEq('get_datetime_minute', dt.minute, 30);
        checkEq('get_datetime_second', dt.second, 0);
    }

    // -- readTemperature(): positive and negative -------------------------
    {
        const conn = new I2CConnectionMock();
        conn.setRegister(_REG_TEMP_MSB, [0x19, 0x40]); // 25 + 0.25 = 25.25
        const rtc = new DS3231Minimal(conn);
        checkTrue('read_temperature_positive', Math.abs((await rtc.readTemperature()) - 25.25) < 1e-9);
    }
    {
        const conn = new I2CConnectionMock();
        conn.setRegister(_REG_TEMP_MSB, [0xFF, 0xC0]); // -1 + 0.75 = -0.25
        const rtc = new DS3231Minimal(conn);
        checkTrue('read_temperature_negative', Math.abs((await rtc.readTemperature()) - (-0.25)) < 1e-9);
    }

    // -- setAlarm1() / getAlarm1() round trip -----------------------------
    {
        const conn = new I2CConnectionMock();
        const rtc = new DS3231Full(conn);
        await rtc.setAlarm1(30, 15, 10, 3, false, DS3231Full.ALARM1_MATCH_HOURS_MINUTES_SECONDS);
        checkEq('alarm1_daydate_reg', conn.registers.get(_REG_ALARM1_DAY_DATE), 0x83);
        const a1 = await rtc.getAlarm1();
        checkEq('alarm1_second', a1.second, 30);
        checkEq('alarm1_minute', a1.minute, 15);
        checkEq('alarm1_hour', a1.hour, 10);
        checkEq('alarm1_day_or_date', a1.dayOrDate, 3);
        checkEq('alarm1_is_day_of_week', a1.isDayOfWeek, false);
        checkEq('alarm1_match_mode', a1.matchMode, DS3231Full.ALARM1_MATCH_HOURS_MINUTES_SECONDS);
    }

    // -- setAlarm2() / getAlarm2() round trip -----------------------------
    {
        const conn = new I2CConnectionMock();
        const rtc = new DS3231Full(conn);
        await rtc.setAlarm2(45, 20, 5, true, DS3231Full.ALARM2_MATCH_HOURS_MINUTES);
        checkEq('alarm2_daydate_reg', conn.registers.get(_REG_ALARM2_DAY_DATE), 0xC5);
        const a2 = await rtc.getAlarm2();
        checkEq('alarm2_minute', a2.minute, 45);
        checkEq('alarm2_hour', a2.hour, 20);
        checkEq('alarm2_day_or_date', a2.dayOrDate, 5);
        checkEq('alarm2_is_day_of_week', a2.isDayOfWeek, true);
        checkEq('alarm2_match_mode', a2.matchMode, DS3231Full.ALARM2_MATCH_HOURS_MINUTES);
    }

    // -- square wave / 32kHz -----------------------------------------------
    {
        const conn = new I2CConnectionMock();
        const rtc = new DS3231Full(conn);
        await rtc.enableSquareWave(4096, true);
        checkEq('enable_square_wave_ctrl', conn.registers.get(_REG_CONTROL), 0x50); // RS2=0x10 | BBSQW=0x40
        await rtc.disableSquareWave();
        checkEq('disable_square_wave_sets_intcn', conn.registers.get(_REG_CONTROL), 0x54);

        checkTrue('rate_1hz_rejects_invalid', await (async () => {
            try { await rtc.enableSquareWave(3000); return false; } catch (e) { return true; }
        })());
    }
    {
        const conn = new I2CConnectionMock();
        conn.setRegister(_REG_CONTROL_STATUS, [0x00]);
        const rtc = new DS3231Full(conn);
        checkEq('is_32khz_enabled_false', await rtc.is32khzEnabled(), false);
        await rtc.enable32khzOutput();
        checkEq('enable_32khz_sets_bit', conn.registers.get(_REG_CONTROL_STATUS) & 0x08, 0x08);
        await rtc.disable32khzOutput();
        checkEq('disable_32khz_clears_bit', conn.registers.get(_REG_CONTROL_STATUS) & 0x08, 0x00);
    }

    // -- oscillator control --------------------------------------------------
    {
        const conn = new I2CConnectionMock();
        conn.setRegister(_REG_CONTROL_STATUS, [0x80]);
        const rtc = new DS3231Full(conn);
        checkEq('oscillator_stopped_true', await rtc.oscillatorStopped(), true);
        await rtc.clearOscillatorStopped();
        checkEq('clear_oscillator_stopped', conn.registers.get(_REG_CONTROL_STATUS) & 0x80, 0x00);

        await rtc.disableBatteryOscillator();
        checkEq('disable_battery_oscillator_sets_eosc', conn.registers.get(_REG_CONTROL) & 0x80, 0x80);
        await rtc.enableBatteryOscillator();
        checkEq('enable_battery_oscillator_clears_eosc', conn.registers.get(_REG_CONTROL) & 0x80, 0x00);
    }

    // -- forced temperature conversion (BSY already clear in the mock) -----
    {
        const conn = new I2CConnectionMock();
        conn.setRegister(_REG_TEMP_MSB, [0x1B, 0x00]); // 27.0 C
        const rtc = new DS3231Full(conn);
        const t = await rtc.forceTemperatureConversion();
        checkEq('force_conversion_sets_conv_bit', conn.registers.get(_REG_CONTROL) & 0x20, 0x20);
        checkTrue('force_conversion_returns_temperature', Math.abs(t - 27.0) < 1e-9);
    }

    // -- aging offset ---------------------------------------------------------
    {
        const conn = new I2CConnectionMock();
        conn.setRegister(_REG_AGING_OFFSET, [0xFF]);
        const rtc = new DS3231Full(conn);
        checkEq('get_aging_offset_negative', await rtc.getAgingOffset(), -1);
        await rtc.setAgingOffset(-5);
        checkEq('set_aging_offset', conn.registers.get(_REG_AGING_OFFSET), 0xFB);
    }

    // -- interrupt API: pollInterrupt / enable / disable ----------------------
    {
        const conn = new I2CConnectionMock();
        conn.setRegister(_REG_CONTROL_STATUS, [0x83]); // OSF=1, A2F=1, A1F=1
        const rtc = new DS3231Full(conn);
        const status = await rtc.pollInterrupt();
        checkEq('poll_interrupt_returns_sources', status, 0x03);
        checkEq('poll_interrupt_clears_only_alarm_flags', conn.registers.get(_REG_CONTROL_STATUS), 0x80);
    }
    {
        const conn = new I2CConnectionMock();
        const rtc = new DS3231Full(conn);
        await rtc.enableInterrupt(DS3231Full.SOURCE_ALARM1 | DS3231Full.SOURCE_ALARM2);
        checkEq('enable_interrupt_sets_a1ie_a2ie_intcn', conn.registers.get(_REG_CONTROL), 0x07);
        await rtc.disableInterrupt(DS3231Full.SOURCE_ALARM1);
        checkEq('disable_interrupt_clears_a1ie_only', conn.registers.get(_REG_CONTROL), 0x06);
    }

    // -- interrupt API: onInterrupt()/offInterrupt() polling fallback --------
    {
        const conn = new I2CConnectionMock();
        conn.setRegister(_REG_CONTROL_STATUS, [0x01]); // A1F latched
        const rtc = new DS3231Full(conn);
        let calls = 0;
        let lastStatus = 0;
        await rtc.onInterrupt((status) => { calls++; lastStatus = status; });
        await new Promise((r) => setTimeout(r, 30));
        await rtc.offInterrupt();
        checkTrue('on_interrupt_fires_once_for_latched_flag', calls === 1);
        checkEq('on_interrupt_callback_status', lastStatus, 0x01);
    }

    console.log(`===DONE: ${passed} passed, ${failed} failed===`);
    process.exit(failed === 0 ? 0 : 1);
}

main();
